flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBConnectorFullChangelogDeserializationSchema.java

```diff
@@ -0,0 +1,820 @@
+import org.apache.kafka.connect.source.SourceRecord;
+import org.bson.BsonBinary;
+import org.bson.BsonBinarySubType;
+import org.bson.BsonDateTime;
+import org.bson.BsonDocument;
+import org.bson.BsonMaxKey;
+import org.bson.BsonMinKey;
+import org.bson.BsonRegularExpression;
+import org.bson.BsonTimestamp;
+import org.bson.BsonUndefined;
+import org.bson.BsonValue;
+import org.bson.codecs.BsonArrayCodec;
+import org.bson.codecs.EncoderContext;
+import org.bson.json.JsonWriter;
+import org.bson.types.Decimal128;
+
+import java.io.Serializable;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.lang.reflect.Array;
+import java.math.BigDecimal;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.time.format.DateTimeFormatter.ISO_OFFSET_DATE_TIME;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Deserialization schema from Mongodb ChangeStreamDocument to Flink Table/SQL internal data that
+ * produces full changelog mode structure {@link RowData}.
+ */
+public class MongoDBConnectorFullChangelogDeserializationSchema
+        implements DebeziumDeserializationSchema<RowData> {
+
+    private static final long serialVersionUID = 1750787080613035184L;
+
+    /** TypeInformation of the produced {@link RowData}. */
+    private final TypeInformation<RowData> resultTypeInfo;
+
+    /** Local Time zone. */
+    private final ZoneId localTimeZone;
+
+    /**
+     * Runtime converter that converts {@link
+     * com.mongodb.client.model.changestream.ChangeStreamDocument}s into {@link RowData} consisted
+     * of physical column values.
+     */
+    protected final DeserializationRuntimeConverter physicalConverter;
+
+    /** Whether the deserializer needs to handle metadata columns. */
+    protected final boolean hasMetadata;
+
+    /**
+     * A wrapped output collector which is used to append metadata columns after physical columns.
+     */
+    private final AppendMetadataCollector appendMetadataCollector;
+
+    public MongoDBConnectorFullChangelogDeserializationSchema(
+            RowType physicalDataType,
+            MetadataConverter[] metadataConverters,
+            TypeInformation<RowData> resultTypeInfo,
+            ZoneId localTimeZone) {
+        this.hasMetadata = checkNotNull(metadataConverters).length > 0;
+        this.appendMetadataCollector = new AppendMetadataCollector(metadataConverters);
+        this.physicalConverter = createConverter(physicalDataType);
+        this.resultTypeInfo = resultTypeInfo;
+        this.localTimeZone = localTimeZone;
+    }
+
+    @Override
+    public void deserialize(SourceRecord record, Collector<RowData> out) throws Exception {
```

Comment:
We can let `MongoDBConnectorFullChangelogDeserializationSchema` extends `MongoDBConnectorDeserializationSchema`  and then  overrides this method to reuse common code pice.